### PR TITLE
Nice-ified `JSONDecodeError`

### PR DESCRIPTION
**Pull request by jamesbraza**
# Description

- Nice-ified `JSONDecodeError` for LLM responses that didn't listen
- Combined two test cases into one (extra parameterized case

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama_index/output_parsers/selection.py [hunk 1]
```diff
@@ -82,8 +82,13 @@ def _format_output(self, output: List[dict]) -> List[dict]:
         return output_json
 
     def parse(self, output: str) -> Any:
-        output = _marshal_llm_to_json(output)
-        json_output = json.loads(output)
+        json_output = _marshal_llm_to_json(output)
```
Comment (jamesbraza) [inline llama_index/output_parsers/selection.py:85]: Note: this also fixes a latent issue, where the returned `StructuredOutput` had a `raw_output` that was already marshalled away from raw

File: llama_index/output_parsers/selection.py [hunk 2]
```diff
@@ -82,8 +82,13 @@ def _format_output(self, output: List[dict]) -> List[dict]:
...
+        try:
+            json_output = json.loads(json_output)
+        except json.decoder.JSONDecodeError as exc:
+            raise ValueError(
+                f"Failed to convert LLM output {output!r} to JSON."
```
Comment (logan-markewich) [inline llama_index/output_parsers/selection.py:90]: huh, first I've seen this syntax -- is this supported by python3.8?

**Review by logan-markewich — APPROVED** [review]

File: llama_index/output_parsers/selection.py (hunk 2 above)
Comment (jamesbraza) [reply llama_index/output_parsers/selection.py:89]: Yeah it's a fun shorthand for `repr`: https://stackoverflow.com/questions/44800801/in-python-format-f-string-strings-what-does-r-mean

There is also `!s`, fwiw

Comment (jamesbraza) [reply llama_index/output_parsers/selection.py:89]: Oh and yeah it works in 3.8

**Comment by jamesbraza** [comment]
@logan-markewich CI passed now, fyi

**Pull request merged**